### ISSUE TITLE
'NoneType' object is not subscriptable error in heterogeneous graph 

### ISSUE THREAD

**Issue by lmnt** [pyg-team/pytorch_geometric#7740]
### 🐛 Describe the bug

```python
import torch
import torch_geometric
import torch.nn as nn
import torch_geometric.nn as gnn
import torch.nn.functional as F

# Define your link prediction model
class GCN(nn.Module):
    def __init__(self, hidden_dim):
        super(GCN, self).__init__()
        self.conv1 = gnn.SAGEConv(hidden_dim, hidden_dim)
        self.conv2 = gnn.SAGEConv(hidden_dim, hidden_dim)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index)
        x = torch.relu(x)
        x = self.conv2(x, edge_index)
        x = torch.relu(x)
        return x

class Model(nn.Module):
    def __init__(self, input_dim, hidden_channels, data):
        super().__init__()
        self.a_lin = torch.nn.Linear(40, hidden_channels)
        self.a_emb = torch.nn.Embedding(data["a"].num_nodes, hidden_channels)
        self.b_emb = torch.nn.Embedding(data["b"].num_nodes, hidden_channels)
        self.gnn = GCN(hidden_channels)
        self.gnn = to_hetero(self.gnn, metadata=data.metadata())
        self.classifier = Classifier(hidden_channels*2) #256+256
    def forward(self, dat):
        # `x_dict` holds feature matrices of all node types
        # `edge_index_dict` holds all edge indices of all edge types
        edge_types = dat.metadata()[1]

        x_dict = {
            "a": self.a_lin(dat["a"].x) + self.a_emb(dat["a"].node_id),
            "b": self.b_emb(dat["b"].node_id),
        }

        x_dict = self.gnn(x_dict, dat.edge_index_dict)

for ix, d in enumerate(train_loader):
    optimizer.zero_grad()    
    output = model(d)
```
Results in the following error: 
```
  File "/opt/conda/lib/python3.10/site-packages/torch/fx/graph_module.py", line 267, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "<eval_with_key>.33", line 8, in forward
    getitem__a = dat__a['a']
TypeError: 'NoneType' object is not subscriptable

Call using an FX-traced Module, line 8 of the traced Module's generated forward function:
def forward(self, dat):
    dat_dict = torch_geometric_nn_to_hetero_transformer_get_dict(dat);  dat = None
    dat__a = dat_dict.get('a', None)
    dat__b = dat_dict.get('b', None);  dat_dict = None
    getitem__a = dat__a['a']

    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    getitem_2__b = dat__b['b']

    getattr_2__a = getitem_2__a.x;  getitem_2__a = None

TypeError                                 Traceback (most recent call last)
Cell In[66], line 11
      9 optimizer.zero_grad()
---> 11 output = model(d)

File /opt/conda/lib/python3.10/site-packages/torch/fx/graph_module.py:658, in GraphModule.recompile.<locals>.call_wrapped(self, *args, **kwargs)
    657 def call_wrapped(self, *args, **kwargs):
--> 658     return self._wrapped_call(self, *args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/torch/fx/graph_module.py:275, in _WrappedCall.__call__(self, obj, *args, **kwargs)
    272 if "eval_with_key" in topmost_framesummary.filename:
    273     print(_WrappedCall._generate_error_message(topmost_framesummary),
    274           file=sys.stderr)
--> 275     raise e.with_traceback(None)
    276 else:
    277     raise e

TypeError: 'NoneType' object is not subscriptable
```

Even though the `dat.metadata` gives:
```
(['a', 'b'],
 [('a', 'associated_with', 'b'),
  ('a', 'associated_with', 'a'),
  ('b', 'rev_associated_with', 'a')])
```

Also, when I try running  the generated forward code manually:
```python
dat_dict = torch_geometric.nn.to_hetero_transformer.get_dict(d)
print(dat_dict.get('a', None), dat_dict["a"])
> None, {'node_id': tensor([    0,     1,     2,  ...}
```

So seems the way that the model is accessing the `HeteroData` object isn't right? Also confused why, once it has accessed the "a" nodes, that it would try to access it again?
i.e. 
```
     dat__a = dat_dict.get('a', None) # get's "a" nodes
     getitem__a = dat__a['a'] # accesses "a" again, even if this error is fixed, this wouldn't work?
```

### Environment

* PyG version: 2.3.0
* PyTorch version: 1.13.0
* OS: Linux
* Python version: 3.10.10
* CUDA/cuDNN version: 11.7
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip
* Any other relevant information (*e.g.*, version of `torch-scatter`):
  torch-cluster==1.6.0 
  torch-scatter==2.1.0 
  torch-sparse==0.6.17 
  torch-spline-conv==1.2.1

**Comment by akihironitta** [comment]
Hi @lmnt sorry for replying very late. I'm happy to help if you could share a complete script before someone who has an idea leaves a comment here :)

**Comment by rusty1s** [comment]
Mh, this is indeed strange - I can't spot any obvious issue here, so a reproducible script would be amazing.

**Comment by lmnt** [comment]
Hi! I do have an update that fixed this. 
I realized when I removed `to_hetero(model)` and only had the `to_hetero(self.gnn)` in the model creation itself, it worked fine. I think doing it twice caused this issue.

Before that, it seemed to want the data input to the model be of the form:
`{"a": HeteroDataObject, "b": HeteroDataObject}`
whereas I was passing in data as `HeteroDataObject` not a dict of them. I will close the issue. Thanks!
